sdks/android/Mentat/library/src/main/java/org/mozilla/mentat/RustIterator.java

```diff
@@ -0,0 +1,69 @@
+/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
+ * Copyright 2018 Mozilla
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License. */
+
+package org.mozilla.mentat;
+
+import com.sun.jna.Pointer;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+// Common factored-out code shared by both RelResultIterator and CollResultIterator (and possibly
+// others in the future). This code is a bit error-prone so it's worth avoiding the duplication.
+abstract class RustIterator<E extends RustObject> extends RustObject implements Iterator<E> {
+    // We own this if it is not null!
+    private Pointer nextPointer;
+
+    RustIterator(Pointer pointer) {
+        super(pointer);
+    }
+
+    /** Implement by calling `JNI.INSTANCE.whatever_iter_next(this.validPointer());` */
+    abstract protected Pointer advanceIterator();
+    // Generally should be implemented as `new E(p)`.
+    abstract protected E constructItem(Pointer p);
+
+    private Pointer consumeNextPointer() {
+        if (this.nextPointer == null) {
+            throw new NullPointerException(this.getClass() + " nextPointer already consumed");
+        }
+        Pointer p = this.nextPointer;
+        this.nextPointer = null;
+        return p;
+    }
+
+    @Override
+    public boolean hasNext() {
```

Comment:
Is this semantically allowed by Java iterators?  (I.e., that calling `hasNext()` might invoke `next()`, potentially consuming a resource or whatever?

Reply:
I think so.

1. This object should be the only thing that has access to the iterator pointer, so nobody should be able to tell if we advance it, so long as we appear to not have advanced it
2. Calling hasNext multiple times will not advance it multiple times. We advance it once just to see if we're at the end or not.
3. Calling next() after hasNext has the same behavior as calling it without.

That is to say, in the API this class exposes, hasNext is a pure function, and next is the only impure one.

Under the hood, hasNext may fetch the item that will be returned by next() the next time you call it, but if you didn't call hasNext, next will just do that.